### ISSUE TITLE
Wrong handling of invalid Curve exchange

### ISSUE THREAD

**Issue by fbacchella** [zeromq/jeromq#676]
When a wrong server public key is given in a Curve protected exchange, jeromq go for a wild lool, as it retry connection immediatly. Of course if fails again.

It fails at [CurveServerMechanism.java.processHello](https://github.com/zeromq/jeromq/blob/master/src/main/java/zmq/io/mechanism/curve/CurveServerMechanism.java#L274) :
```
        int rc = cryptoBox.open(helloPlaintext, helloBox, helloBox.capacity(), helloNonce, cnClient, secretKey);
        if (rc != 0) {
            return ZError.EPROTO;
        }
```

The equivalent C code in libzmq [says](https://github.com/zeromq/libzmq/blob/master/src/curve_server.cpp#L188):
```
rc = crypto_box_open (hello_plaintext, hello_box, sizeof hello_box,
                          hello_nonce, _cn_client, _secret_key);
    if (rc != 0) {
        // CURVE I: cannot open client HELLO -- wrong server key?
        session->get_socket ()->event_handshake_failed_protocol (
          session->get_endpoint (), ZMQ_PROTOCOL_ERROR_ZMTP_CRYPTOGRAPHIC);
        errno = EPROTO;
        return -1;
    }
```

So the insertion of ZMQ_PROTOCOL_ERROR_ZMTP_CRYPTOGRAPHIC is lost. I don't know how it should be implemented in jeromq.

**Comment by trevorbernard** [comment]
It appears it was added in this commit of libzmq: https://github.com/zeromq/libzmq/commit/e2d3ba9c6

**Comment by fredoboulo** [comment]
Current implementation is based on https://github.com/zeromq/zeromq4-1, it could be possible to transcript the commit in jeromq as well.

**Comment by fbacchella** [comment]
This one is going to be thought at least for me.
I have been able to make the server return an error by setting:
```
        //  Open Box [64 * %x0](C'->S)
        int rc = cryptoBox.open(helloPlaintext, helloBox, helloBox.capacity(), helloNonce, cnClient, secretKey);
        if (rc != 0) {
            state = State.SEND_ERROR;
            statusCode ="999";
            return 0;
        }
```
 
It's then returned to the client, an error is sent and errno is set to EPROTO latter in:
```
    private boolean processHandshakeCommand(Msg msg)
    {
...
            else if (mechanism.status() == Mechanism.Status.ERROR) {
                errno.set(ZError.EPROTO);
                return false;
            }
```

But when CurveClientMechanism.processError received the error, it just see an error but don't give up the connection and retry it.

Also when I look at SocketBase, I think a lot of event are missing. Even when comparing to the version before patch zeromq/libzmq@e2d3ba9, see 
https://github.com/zeromq/libzmq/commit/e2d3ba9c62c780e0fb5974f7a1dcff7a8d4184c9?diff=split#diff-98bda77577a7f2d4df8770fdf7857bec

That's a part of me problem, but I'm strugling to see the big picture.

**Comment by fbacchella** [comment]
This patch is finally not related to error handling. It just improve event monitoring (I'm preparing a patch for that). The problem is that jeromq (and event libzmq I'm afraid) don't fully implement the RFC.
It says:
```
To reject a client connection explicitly, a server SHALL send an ERROR command:

S:ERROR
```

But jeromq don't do that. It just close the connection. But when I comparge the C an java implementation, I don't see meaningful differences about that. So I think the problem is very deep.

**Comment by fbacchella** [comment]
The missing events should be addressed in #825.
